crypto/ocsp/ocsp_http.c

```diff
@@ -22,11 +26,335 @@
 #define OHS_NOREAD              0x1000
 // Error condition
 #define OHS_ERROR               (0 | OHS_NOREAD)
+// First line being read
+#define OHS_FIRSTLINE           1
+// MIME headers being read
+#define OHS_HEADERS             2
+// OCSP initial header (tag + length) being read
+#define OHS_ASN1_HEADER         3
+// OCSP content octets being read
+#define OHS_ASN1_CONTENT        4
 // First call: ready to start I/O
 #define OHS_ASN1_WRITE_INIT     (5 | OHS_NOREAD)
+// Request being sent
+#define OHS_ASN1_WRITE          (6 | OHS_NOREAD)
+// Request being flushed
+#define OHS_ASN1_FLUSH          (7 | OHS_NOREAD)
+// Completed
+#define OHS_DONE                (8 | OHS_NOREAD)
 // Headers set, no final \r\n included
 #define OHS_HTTP_HEADER         (9 | OHS_NOREAD)
 
+// Parse the HTTP response. This will look like this: "HTTP/1.0 200 OK". We
+// need to obtain the numeric code and (optional) informational message.
+static int parse_http_line(char *line)
+{
+    int http_code;
+    char *code, *reason, *end;
+    // Skip to first white space (passed protocol info)
+    for (code = line; *code != '\0' && !isspace(*code); code++) {
+      continue;
+    }
+    if (*code == '\0') {
+        OPENSSL_PUT_ERROR(OCSP,  OCSP_R_SERVER_RESPONSE_PARSE_ERROR);
+        return 0;
+    }
+
+    // Skip past white space to start of response code.
+    while (*code != '\0' && isspace(*code)) {
+      code++;
+    }
+    if (*code == '\0') {
+        OPENSSL_PUT_ERROR(OCSP,  OCSP_R_SERVER_RESPONSE_PARSE_ERROR);
+        return 0;
+    }
+
+    // Find end of response code: first whitespace after start of code.
+    for (reason = code; *reason != '\0' && !isspace(*reason); reason++) {
+      continue;
+    }
+    if (*reason == '\0') {
+        OPENSSL_PUT_ERROR(OCSP,  OCSP_R_SERVER_RESPONSE_PARSE_ERROR);
+        return 0;
+    }
+
+    // Set end of response code and start of message.
+    *reason++ = '\0';
+    // Attempt to parse numeric code
+    http_code = (int)strtoul(code, &end, 10);
+    if (*end != '\0') {
+      return 0;
+    }
+
+    // Skip over any leading white space in message.
+    while (*reason != '\0' && isspace(*reason)) {
+      reason++;
+    }
+    if (*reason != '\0') {
+        // Finally, zap any trailing white space in message (include CRLF).
+        // We know reason has a non-white space character so this is OK.
+        for (end = reason + strlen(reason) - 1; isspace(*end); end--) {
+          *end = '\0';
+        }
+    }
+    if (http_code != 200) {
+        OPENSSL_PUT_ERROR(OCSP,  OCSP_R_SERVER_RESPONSE_PARSE_ERROR);
+        if (*reason == '\0') {
+          ERR_add_error_data(2, "Code=", code);
+        }
+        else {
+          ERR_add_error_data(4, "Code=", code, ",Reason=", reason);
+        }
+        return 0;
+    }
+    return 1;
+}
+
+int OCSP_REQ_CTX_nbio(OCSP_REQ_CTX *rctx)
```

Comment:
Can you include a high level overview of what this function is doing and why? I see in internal.h above it is "exchanging request and response via HTTP on (non-)blocking BIO in rctx", but this function is doing a lot of work that is not immediately obvious how it related back to the goal of sending and receiving a response.